lib/domain.js

```diff
@@ -48,13 +48,56 @@ Object.defineProperty(process, 'domain', {
   }
 });
 
+const pairing = new Map();
+const asyncHook = async_hooks.createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+
+    if (process.domain !== null && process.domain !== undefined) {
+      // if this operation is created while in a domain, let's mark it
+      pairing.set(asyncId, process.domain);
+      resource.domain = process.domain;
+      if (resource.promise !== undefined &&
+        resource.promise instanceof Promise) {
+        // resource.promise instanceof Promise make sure that the
+        // promise comes from the same context
+        // see https://github.com/nodejs/node/issues/15673
+        resource.promise.domain = process.domain;
+      }
+    }
+  },
+  before(asyncId) {
+
+    if (pairing.has(asyncId)) { // enter domain for this cb
```

Comment:
How much faster is `has()` than `get()`? Would it be worth just calling `get()` and checking for `undefined`?

Reply:
I have no data here, but it would seems to me that accessing the map only once might be more efficient.